pydoover/cloud/reports/application.py

```diff
@@ -0,0 +1,258 @@
+        except Exception as e:
+            log.error(f"Error creating report message: {e}")
+            return
+
+        if "id" not in data:
+            log.error("Error creating report message: cannot find id in response")
+            return
+
+        log.info(f"Message Creation took {time.perf_counter() - s} seconds.")
+
+        self._report_id = data["id"]
+        self.report_id = data["id"]
+
+        log.info(f"Report ID: {self.report_id}")
+
+        await self._generate(self.devices, self.period_start, self.period_end)
+
+    async def on_single_execute(self, event: ScheduleEvent):
+        if len(self.config.report_config.devices.elements) == 0:
+            log.error("No Devices provided")
+            return
+
+        if self.config.report_config.period_start.value is None:
+            log.error("No Period Start provided")
+            return
+
+        if self.config.report_config.period_end.value is None:
+            log.error("No Period End provided")
+            return
+
+        if self.config.report_config.report_id.value is None:
+            log.error("No Report Id provided")
+            return
+
+        devices = [
+            device.value for device in self.config.report_config.devices.elements
+        ]
+        self._generate(
+            devices,
+            self.config.report_config.period_start.value,
+            self.config.report_config.period_end.value,
+        )
+
+    async def _generate(
+        self, agent_ids: list[int], period_start: datetime, period_end: datetime
+    ):
+        log.info(f"Start Report Generation from {period_start} to {period_end}")
+        s = time.perf_counter()
+        
+        files = await self.generate(agent_ids, period_start, period_end)
+        
+        if type(files) is not list:
+            files = [files]
+        
+        log.info(
+            f"Report Generation took {time.perf_counter() - s} seconds. {len(files)} file{'s' if len(files) > 1 else ''} generated."
+        )
+
+        self._report_metadata["status"] = "Complete"
+
+        log.info("Uploading Report to Doover")
+        s = time.perf_counter()
+        try:
+            await self.api.update_message(
+                self.agent_id,
+                "reports",
+                self._report_id,
+                self._report_metadata,
+                organisation_id=self.agent_id,
+                files=files,
+            )
+        except Exception as e:
+            log.error(f"Error creating report message: {e}")
+            return
+        log.info(f"Report Upload took {time.perf_counter() - s} seconds.")
+
+    async def generate(
+        self, agent_ids: list[int], period_start: datetime, period_end: datetime
+    ):
+        """Override this method to specify how a report should be generated.
+
+        You do **not** need to call `super().generate()` as this function ordinarily does nothing.
+        """
+        return NotImplemented
+
+    async def fetch_device_in_window(
```

Comment:
nitpick naming thing, but should this be 
```diff
- async def fetch_device_in_window(
+ async def fetch_device_channel_in_window(
``` 

because it's really the channel we're operating on? I'd even go so far as to say you could drop the `device` because every channel needs a device (agent) and you're not actually checking that the `agent_id` is a device